### PR TITLE
Add profiles and profile-scoped paths

### PR DESCRIPTION
**Pull request by geigerzaehler**
To address #488 we add the `Profile` struct.

This pull request is against `next`. I will try to write automatic migration code for this so we can include it in `maint` without a breaking change.

We’ve also simplified some methods on `Paths` and made them private.

### CODE REVIEW

File: librad/src/profile.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+use uuid::Uuid;
+
+use crate::paths::{project_dirs, Paths};
+
+/// A [`Profile`] provides [`Paths`] scoped by an identifier.
+///
+/// Profiles with different identifiers have distinct paths.
+#[derive(Debug, Clone)]
+pub struct Profile {
+    id: String,
```
Comment (FintanH) [inline librad/src/profile.rs:16]: Would there be any advantage to using the `Uuid` type here instead? Or even having a generic parameter for switching between different identity types in the future or for testing?

File: librad/src/profile.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    paths: Paths,
+}
+
+impl Profile {
+    /// Creates a profile by loading the the profile identifier and paths from
```
Comment (FintanH) [inline librad/src/profile.rs:21]:
```suggestion
    /// Creates a profile by loading the profile identifier and paths from
```

File: librad/src/profile.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    /// the environment variables or well-known file.
+    ///
+    /// If `RAD_PROFILE` is not set, the value is used as the profile identifier
```
Comment (FintanH) [inline librad/src/profile.rs:24]: Should this be, "If `RAD_PROFILE` is set"? Otherwise, what "value" are you referring to?

File: librad/src/paths.rs [hunk 4]
```diff
@@ -72,7 +63,7 @@ impl Paths {
         &self.git_includes_dir
     }
 
-    pub fn all_dirs(&self) -> HashMap<&str, &Path> {
+    fn all_dirs(&self) -> impl Iterator<Item = &Path> {
```
Comment (FintanH) [inline librad/src/paths.rs:66]: What's the reason for this change?

**Review by FintanH — COMMENTED** [review]
Looking good. I just have a few details I'd like to clear up :)

File: librad/src/profile.rs (hunk 1 above)
Comment (geigerzaehler) [reply librad/src/profile.rs:16]: I don’t see a strong need to make this generic. At the moment he tests don’t benefit from it and I can’t imagine that in the future we need a profile with two different ID parameters. If that happens though, it’s not to late to change the code.

Regarding using `Uuid` instead of `String`. If we implement this than only UUIDs are valid profile identifiers. This is a limitation that may make the feature harder to user. For example, say I’m playing around with this using the `RAD_PROFILE` environment variable. Since UUIDs are quite opaque, I’m free to use `RAD_PROFILE=foo` and radicle will deal with this. We only use `UUIDs` to generate unspecified profiles to ensure they are unique.

Comment (FintanH) [reply librad/src/profile.rs:16]: Makes sense :+1:

File: librad/src/paths.rs (hunk 4 above)
Comment (geigerzaehler) [reply librad/src/paths.rs:66]: I’ve removed `pub` because this method is unused by external consumers and keeping the public API small makes it easier to grasp and harder to inadvertently break.

I’ve replaced `HashMap` with the iterator because for consumers it provides all it needs and it simplifies the implementation.

I’m ok with reverting all those secondary changes. They are not related to the functionality but seemed like nice improvements.

File: librad/src/profile.rs (hunk 3 above)
Comment (geigerzaehler) [reply librad/src/profile.rs]: Yes, of course! I’ve fixed this.

File: librad/src/paths.rs [hunk 5]
```diff
@@ -81,22 +72,33 @@ impl Paths {
...
-        self.all_dirs().values().try_for_each(fs::create_dir_all)?;
+        self.all_dirs().try_for_each(fs::create_dir_all)?;
         Ok(self)
     }
 }
 
+/// Returns [`ProjectDirs`] for this specific project (`radicle`).
+///
+/// Returns [`io::Error`] if the project directories could not be determined,
+/// most likely du to the `$HOME` environment variable missing
```
Comment (kim) [inline librad/src/paths.rs:92]:
```suggestion
/// most likely due to the `$HOME` environment variable missing
```

File: librad/src/paths.rs [hunk 6]
```diff
@@ -81,22 +72,33 @@ impl Paths {
...
+pub(crate) fn project_dirs() -> Result<ProjectDirs, io::Error> {
+    ProjectDirs::from("xyz", "radicle", "radicle").ok_or_else(|| {
```
Comment (kim) [inline librad/src/paths.rs:94]: Do we want to take the opportunity to change this to `xyz`, `radicle`, `radicle-link`? I'd think that migration logic would be greatly simplified, too.

File: librad/src/profile.rs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    /// If `RAD_PROFILE` is set, the value is used as the profile identifier
+    /// and system paths, scoped to the identifier, are used. See
+    /// [`Paths::new`] for details.
+    ///
+    /// Otherwise and if the `RAD_HOME` environment variable is set, the value
+    /// is used to construct the profile with [`Profile::from_root`].
+    ///
+    /// Otherwise, the profile identifier is read from a well-known file and
+    /// system paths scoped to the identifier are used. If the file containing
+    /// the identifier  does not exist, it is created and new identifier is
```
Comment (kim) [inline librad/src/profile.rs:33]:
```suggestion
    /// the identifier  does not exist, it is created and a new identifier is
```

File: librad/src/paths.rs (hunk 4 above)
Comment (FintanH) [reply librad/src/paths.rs:66]: Ya, it seems like its uses were just getting the `values` of the `HashMap` anyway, and I believe we usually use the direct accessors in upstream.

**Review by FintanH — DISMISSED** [review]
Looks good to me

File: librad/src/profile.rs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    /// generated and written to the file.
+    ///
+    /// On Linux, the path to the active profile is
+    /// `$XDG_CONFIG_HOME/radicle/active_profile`.
+    pub fn load() -> Result<Self, io::Error> {
```
Comment (kim) [inline librad/src/profile.rs:38]: Might be good to ensure that the profile id is only a path _component_, or otherwise ensure that it doesn't look funny (eg. `../../WORKGROUP`).

Maybe we can also hand out the id as a `Path`, to reduce suprises when it contains newlines or other funky characters.

**Review by kim — COMMENTED** [review]
looks good, only minor remarks

File: librad/src/paths.rs (hunk 6 above)
Comment (geigerzaehler) [reply librad/src/paths.rs:94]: Good idea, since this is a breaking change anyway.

File: librad/src/profile.rs (hunk 8 above)
Comment (geigerzaehler) [reply librad/src/profile.rs:38]: > Might be good to ensure that the profile id is only a path _component_, or otherwise ensure that it doesn't look funny (eg. `../../WORKGROUP`).

Good point. Will add the validation logic.

> Maybe we can also hand out the id as a `Path`, to reduce suprises when it contains newlines or other funky characters.

Are your referring to `Profile::id`? Having it return `Path` might be confusing, especially since it can only be a path segment since the base directories are all different.

Comment (kim) [reply librad/src/profile.rs:38]: Ya I was thinking of the "smart" `Display` impl for `Path`. If we sanitise the input properly, I'm ok with `&str`.

File: librad/src/paths.rs (hunk 6 above)
Comment (geigerzaehler) [reply librad/src/paths.rs]: Addressed in https://github.com/radicle-dev/radicle-link/pull/496/commits/79a81a3177810bf278b8ba73e644bd822aed573f

File: librad/src/profile.rs (hunk 8 above)
Comment (geigerzaehler) [reply librad/src/profile.rs]: Addressed in https://github.com/radicle-dev/radicle-link/pull/496/commits/79a81a3177810bf278b8ba73e644bd822aed573f

File: librad/src/profile.rs [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    path,
+    path::{Path, PathBuf},
+};
+use uuid::Uuid;
+
+use crate::paths::{project_dirs, Paths};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Invalid profile ID in RAD_PROFILE environment variable: {id}")]
```
Comment (FintanH) [inline librad/src/profile.rs:19]: The convention for error messages is starting with lowercase, as mentioned in `radicle-decisions`

File: librad/src/profile.rs [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+/// Returns `true` if `id` is a valid profile ID.
+///
+/// A valid profile ID is not empty, does not contain path separators, is not
+/// a windows path prefix like `C:`, and is not a special component like `.` or
+/// `..`.
+fn is_valid_profile_id(id: &str) -> bool {
+    let mut components = Path::new(id).components();
+    if let Some(c) = components.next() {
```
Comment (FintanH) [inline librad/src/profile.rs:103]: I think this could be cleaned up to:
```rust
match components.next() {
  Some(path::Component::Normal(_)) => {},
  _ => return false,
}
```

File: librad/src/profile.rs [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+/// Read the profile ID from a file or create the file and write a newly
+/// generated profile ID to it.
+fn load_profile_id(config_dir: &Path) -> Result<String, Error> {
+    let active_profile_path = config_dir.join("active_profile");
+
+    match fs::read_to_string(&active_profile_path) {
+        Ok(content) => {
+            let id = content.lines().next().unwrap();
```
Comment (FintanH) [inline librad/src/profile.rs:126]: Is this `unwrap` safe?

File: librad/src/profile.rs [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+    path::{Path, PathBuf},
+};
+use uuid::Uuid;
+
+use crate::paths::{project_dirs, Paths};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Invalid profile ID in RAD_PROFILE environment variable: {id}")]
+    InvalidProfileIdFromEnv { id: String },
```
Comment (FintanH) [inline librad/src/profile.rs:20]: Is this used? :thinking:

File: librad/src/profile.rs (hunk 11 above)
Comment (geigerzaehler) [reply librad/src/profile.rs:126]: Yes. Will add a comment. Or is the preferred way to put the explanation as the argument to `expect`?

File: librad/src/profile.rs (hunk 12 above)
Comment (geigerzaehler) [reply librad/src/profile.rs:20]: Oops, forgot about this.

File: librad/src/profile.rs [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+        if let Ok(id) = env::var("RAD_PROFILE") {
+            if !is_valid_profile_id(&id) {
+                return Err(Error::InvalidProfileIdFromEnv { id });
+            }
+            let paths = Paths::new(&id)?;
+            return Ok(Self { id, paths });
+        }
+
+        if let Ok(rad_home) = env::var("RAD_HOME") {
+            return Self::from_root(Path::new(&rad_home));
```
Comment (MeBrei) [inline librad/src/profile.rs:68]: Shouldn't the `Paths` still contain the `profile id` here? Otherwise if you set `RAD_HOME` (eg when starting upstream) you wont have the multiple profile feature

File: librad/src/profile.rs (hunk 11 above)
Comment (FintanH) [reply librad/src/profile.rs]: Right, I'd have expected the contents would need to be non-empty :) Sure the file could be there but what happens if it's empty :smile:

Comment (MeBrei) [reply librad/src/profile.rs]: I tested this and it threw an error for me. Maybe I removed the profile id in an incorrect way, but it doesn't seem safe to assume it being removed in a certain way. (What I did: opened the file with `vim`, used `dd` to remove the line, closed with `:x`)

File: librad/src/profile.rs (hunk 13 above)
Comment (geigerzaehler) [reply librad/src/profile.rs:68]: Good point. I was thinking about `RAD_HOME` as an exstension of profile selection where it would not only allow you to control the profile ID (like `RAD_PROFILE`) but also the location of all the data.

What you are suggesting would be to have `RAD_HOME` behave like setting all `XDG_*_HOME` variables all at once and having profiles work just like without `RAD_HOME` but under a different base directory.

I think this is the better approach as it decouples `RAD_HOME` and profiles and is a bit more predictable. @FintanH, @kim any thoughts on this?

Comment (FintanH) [reply librad/src/profile.rs:68]: I'm not understanding the conversation here so I'll try and see how far my understanding goes.

So `RAD_HOME` decides where the root is which could contain multiple profiles and the active profile, i.e.
```
RAD_HOME
|--- active_profile
|--- 1
|    |--- git
|    |--- etc.
|--- 2
|    |---- git
```

But what @MeBrei is suggesting that `RAD_HOME` would point to _one_ of those profiles, for example, `RAD_HOME=<ROOT>/2`, is that correct?

Comment (MeBrei) [reply librad/src/profile.rs:68]: > So `RAD_HOME` decides where the root is which could contain multiple profiles and the active profile, i.e.
> 
> ```
> RAD_HOME
> |--- active_profile
> |--- 1
> |    |--- git
> |    |--- etc.
> |--- 2
> |    |---- git
> ```
> 
@FintanH the above is what I was suggesting. 
Below is what is currently happening, if you set `RAD_HOME` to a profile directory, but it in my tests I set it to a non-profile directory and it resulted in a mixed version, where the proxy store was under a profile id, but the librad directories weren't
> But what @MeBrei is suggesting that `RAD_HOME` would point to _one_ of those profiles, for example, `RAD_HOME=<ROOT>/2`, is that correct?

File: librad/src/profile.rs [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+
+        let profile_id_path = tempdir.path().join("active_profile");
+        let content = "foo\n";
+        std::fs::write(profile_id_path, content).unwrap();
+        let id = load_profile_id(tempdir.path()).unwrap();
+        assert_eq!(id, "foo");
+    }
+
+    #[test]
+    fn empty_profile() {
```
Comment (FintanH) [inline librad/src/profile.rs:202]: This doesn't seem to be testing the empty case, it looks like the test case above? :eyes:

File: librad/src/paths.rs [hunk 15]
```diff
@@ -40,9 +41,7 @@ impl Paths {
         .init()
     }
 
-    // Don't use system paths, but the supplied directory as a root.
-    //
-    // For testing, you know.
+    /// All paths are contained in the given directory.
     pub fn from_root(root: impl AsRef<Path>) -> Result<Self, io::Error> {
```
Comment (FintanH) [inline librad/src/paths.rs:45]: In relation to https://github.com/radicle-dev/radicle-link/pull/496/files#r565931339, this doesn't take into account the `id`, so this https://github.com/radicle-dev/radicle-link/pull/496/files#diff-b5c21e30928026061fdc46f4a876c7bc8374ebe78fe68851375632dffe20d472R83 line seems off to me. You're creating the id, but not using it.

File: librad/src/profile.rs (hunk 13 above)
Comment (FintanH) [reply librad/src/profile.rs:68]: Mmmm I think I'm piecing it together now. Something seems off to me which I commented here https://github.com/radicle-dev/radicle-link/pull/496/files#r566020184

File: librad/src/profile.rs (hunk 11 above)
Comment (geigerzaehler) [reply librad/src/profile.rs]: I think it makes sense to treat an empty (but present) `active_profile` file like there is no ID yet. Will update the PR.

Comment (FintanH) [reply librad/src/profile.rs]: What kind of scenario would that be useful for?

Comment (geigerzaehler) [reply librad/src/profile.rs]: I think it’s just to avoid confusion on the users side. I can definitely foresee somebody deleting the content of the file and then wondering why profile is invalid. If you feel strongly otherwise, @FintanH, we can also leave as it is.

File: librad/src/paths.rs (hunk 15 above)
Comment (geigerzaehler) [reply librad/src/paths.rs:45]: You’re right, it does not take the profile ID into account. We could add the profile ID as a second argument and join it to root. But I’d rather prefer to do that at the call site, that is `Profile::from_root`. The latter isn’t implemented yet as @MeBrei  correctly pointed out.

File: librad/src/profile.rs (hunk 13 above)
Comment (geigerzaehler) [reply librad/src/profile.rs:68]: I’ve updated the PR (and added test) for to implement multi-profile support within `RAD_HOME`. I hope this should clarify things.

File: librad/src/profile.rs (hunk 11 above)
Comment (MeBrei) [reply librad/src/profile.rs]: The idea is to trigger a new Identity being created when no `active_profile` can be found. I think we should serve to both cases if the user removes the contents of the file or deletes the file completely.

Comment (FintanH) [reply librad/src/profile.rs]: Ya, I would think this is an error variant. I wouldn't think that an empty profile is a valid one.

File: librad/src/profile.rs (hunk 13 above)
Comment (kim) [reply librad/src/profile.rs:68]: I'm good with what's in c8e6f5d43c8ac8a94e5ba3039420a147e19a4a63.

Comment (geigerzaehler) [reply librad/src/profile.rs:68]: Sorry folks, didn’t actually push the updates. Could you have another look?

File: librad/src/profile.rs (hunk 11 above)
Comment (geigerzaehler) [reply librad/src/profile.rs]: There seems to be some disagreement here, @MeBrei, @FintanH. Either treat an empty file as an error or treat it as no active profile does exist.

After giving this some thought I believe it would be best to treat an empty file as an error. This makes the behavior more straight-forward and reduces the number of edge cases. If the user deletes the content of the file then we throw an error instead of working around this.

Comment (FintanH) [reply librad/src/profile.rs]: Makes sense to me :ok_hand:

File: librad/src/profile.rs (hunk 13 above)
Comment (kim) [reply librad/src/profile.rs:68]: Hm ok, so `RAD_HOME` and `RAD_PROFILE` are effectively mutually exclusive -- if I set `RAD_PROFILE`, I'd need to set `XDG_*_HOME` to make data go into a non-standard location (which is only possible under Linux). If I set `RAD_HOME`, the only way to control profile selection is via the `active_profile` file, which must already exist.

I find this _more_ confusing, but we all seem to have different expectations there. Perhaps we should continue to consider `RAD_HOME` as a convenience for test setups (not "stable API"), and look into respecting `XDG_*` variables regardless of platform?

Comment (MeBrei) [reply librad/src/profile.rs:68]: I also assumed both env variables could be combined. We already "released" `RAD_HOME` as a possibility to have an independent dev setup in upstream.

Comment (geigerzaehler) [reply librad/src/profile.rs]: As discussed in https://github.com/radicle-dev/radicle-link/issues/488#issuecomment-770923997 we will go with @MeBrei’s suggestion. This is implemented in f36bfc9a8aad7acdc637f5abd913a1c4dba61915

File: librad/src/profile.rs (hunk 14 above)
Comment (geigerzaehler) [reply librad/src/profile.rs]: Fxied

File: librad/src/paths.rs (hunk 15 above)
Comment (geigerzaehler) [reply librad/src/paths.rs:45]: Fixed in f36bfc9a8aad7acdc637f5abd913a1c4dba61915

**Comment by geigerzaehler** [comment]
Thanks for all the input @kim, @FintanH, @MeBrei. Should be ready for review again.

**Review by kim — APPROVED** [review]
:+1:

Do we not need a persistent `set_active_profile` yet?

File: librad/src/profile.rs [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+
+impl Profile {
+    /// Creates a profile by loading the profile identifier and paths from
+    /// the environment variables or well-known file.
+    ///
+    /// By default, the profile identifier is read from the `active_profile`
+    /// file in a system specific location. The paths returned by
+    /// [`Profile::paths`] are based on system paths (see [`directories::
+    /// ProjectDirs`]) and include the profile identifier. If the file
+    /// containing the identifier  does not exist, it is created and a new
```
Comment (FintanH) [inline librad/src/profile.rs:44]:
```suggestion
    /// containing the identifier does not exist, it is created and a new
```

**Comment by geigerzaehler** [comment]
> Do we not need a persistent `set_active_profile` yet?

Not yet. In upstream you can only use `RAD_PROFILE` to switch between profiles.

File: librad/src/profile.rs [hunk 17]
```diff
@@ -37,51 +37,68 @@ impl Profile {
...
         }
 
         if let Ok(rad_home) = env::var("RAD_HOME") {
-            return Self::from_root(Path::new(&rad_home));
-        }
-
-        let id = load_profile_id(project_dirs()?.config_dir())?;
+            Self::from_root(Path::new(&rad_home), env_profile_id)
+        } else {
+            let id = if let Some(id) = env_profile_id {
```
Comment (FintanH) [inline librad/src/profile.rs:74]: This suspiciously looks like `from_root` :eyes: Would I be right in thinking that this is `Self::from_root(project_dirs()?.config_dir()), env_profile_id)`? And then ultimately these just differ on the `root` being passed, i.e. `RAD_HOME` or the project dirs

File: librad/src/profile.rs [hunk 18]
```diff
@@ -219,15 +231,4 @@ mod test {
         let id = load_profile_id(tempdir.path()).unwrap();
         assert_eq!(id, "foo");
     }
-
-    #[test]
-    fn empty_profile_file() {
```
Comment (FintanH) [inline librad/src/profile.rs]: Wait, is there a test for the empty file causing an error?

**Review by FintanH — COMMENTED** [review]
Looks good :+1: Few nits/questions

File: librad/src/profile.rs (hunk 17 above)
Comment (geigerzaehler) [reply librad/src/profile.rs:74]: The difference between this and `Profile::from_root` is that here we call `Paths::new(id)` whereas in `Profile::from_root` we call `Paths::from_root(..)`; and both `Paths` constructors also differ. What this has in common with `Profile::from_root` is unwrapping the profile ID or loading it from file, but that’s not too much and doesn’t warrant extracting it.

File: librad/src/profile.rs (hunk 18 above)
Comment (geigerzaehler) [reply librad/src/profile.rs]: Added in https://github.com/radicle-dev/radicle-link/pull/496/commits/d33410cc76657db918e3ae259ee00623acaf1741

File: librad/src/profile.rs (hunk 17 above)
Comment (FintanH) [reply librad/src/profile.rs:74]: Ah you're right. Thanks for checking that :)

**Review by FintanH — APPROVED** [review]
Great! Thanks for the good work here. I'm glad we were able to work out the finer details :)
